### PR TITLE
feat(connector): implement Authorize for mollie klarna

### PR DESCRIPTION
**Pull request by shuklatushar226**
## Summary

Implement **Authorize** flow for **Mollie** connector (Klarna / PayLater payment method).

This implementation was generated and validated by **GRACE** (automated connector integration pipeline).

## Changes

- Added Klarna (PayLater) support to the Authorize flow in `mollie/transformers.rs`
- Added `MolliePaymentMethodData::Klarna` variant plus `KlarnaMethodData`, `MollieKlarnaAddress`, and `MollieLineItem` request types
- Added `TryFrom` handling for `PaymentMethodData::PayLater(PayLaterData::KlarnaRedirect {})`, building the Mollie billing address and order lines

## Files Modified

- crates/integrations/connector-integration/src/connectors/mollie/transformers.rs

## gRPC Test Results

**Status: PASS**

<details>
<summary>grpcurl Authorize call (credentials redacted)</summary>

```
--- grpcurl: Authorize (Klarna / PayLater) ---
(server run on ports 18000/18080 via CS__SERVER__PORT/CS__METRICS__PORT because default ports 8000/8080 were held by another process)

grpcurl -plaintext \
  -H 'x-connector: mollie' \
  -H 'x-auth: body-key' \
  -H 'x-api-key: <REDACTED>' \
  -H 'x-key1: <REDACTED>' \
  -d '{
    "merchant_transaction_id": "test_mollie_klarna_001",
    "amount": {"minor_amount": 1000, "currency": "EUR"},
    "payment_method": {"klarna": {}},
    "capture_method": "AUTOMATIC",
    "address": {"billing_address": {
      "first_name": {"value": "John"}, "last_name": {"value": "Doe"},
      "line1": {"value": "<REDACTED>"}, "city": {"value": "Amsterdam"},
      "zip_code": {"value": "<REDACTED>"}, "country_alpha2_code": "NL",
      "email": {"value": "<REDACTED>"}}},
    "auth_type": "NO_THREE_DS",
    "return_url": "https://example.com/return",
    "description": "Klarna pay later test"
  }' \
  localhost:18000 types.PaymentService/Authorize

Response:
{
  "merchantTransactionId": "tr_kaSK3R5vk9QNR2Jrk3pSJ",
  "connectorTransactionId": "tr_kaSK3R5vk9QNR2Jrk3pSJ",
  "status": "AUTHENTICATION_PENDING",
  "statusCode": 201,
  "redirectionData": {
    "form": {
      "endpoint": "https://www.mollie.com/checkout/test-mode",
      "method": "HTTP_METHOD_GET",
      "formFields": {"method": "klarna", "token": "7.i8iu20"}
    }
  }
}

Outgoing request to Mollie confirmed: POST https://api.mollie.com/v2/payments with method "klarna", billingAddress, and lines. Mollie responded 201 status "open" with checkout URL.
```

</details>

<details>
<summary>grpcurl Authorize call — latest verified run (credentials redacted)</summary>

```
--- grpcurl: Authorize (Klarna / PayLater) ---

grpcurl -plaintext \
  -H 'x-connector: mollie' \
  -H 'x-auth: body-key' \
  -H 'x-api-key: <REDACTED>' \
  -H 'x-key1: <REDACTED>' \
  -d '{
    "merchant_transaction_id": "test_mollie_klarna_001",
    "amount": {"minor_amount": 1000, "currency": "EUR"},
    "payment_method": {"klarna": {}},
    "capture_method": "AUTOMATIC",
    "address": {"billing_address": {
      "first_name": {"value": "John"}, "last_name": {"value": "Doe"},
      "line1": {"value": "Keizersgracht 126"}, "city": {"value": "Amsterdam"},
      "zip_code": {"value": "1015 CW"}, "country_alpha2_code": "NL",
      "email": {"value": "test@example.com"}}},
    "auth_type": "NO_THREE_DS",
    "return_url": "https://example.com/return",
    "description": "Klarna pay later test"
  }' \
  localhost:18000 types.PaymentService/Authorize

Response:
{
  "merchantTransactionId": "tr_zwTXhJQCmxnKrGS47dpSJ",
  "connectorTransactionId": "tr_zwTXhJQCmxnKrGS47dpSJ",
  "status": "AUTHENTICATION_PENDING",
  "statusCode": 201,
  "redirectionData": {
    "form": {
      "endpoint": "https://www.mollie.com/checkout/test-mode",
      "method": "HTTP_METHOD_GET",
      "formFields": {
        "method": "klarna",
        "token": "7.2x6b8s"
      }
    }
  }
}
```

</details>

## Validation Checklist

- [x] `cargo build` passed with zero errors
- [x] grpcurl Authorize returned success status (2xx)
- [x] No credentials in committed source code
- [x] Only connector-specific files modified

### CODE REVIEW

**Comment by XyneSpaces** [comment]
## Automated Review - Mollie Klarna Implementation (PR #1579)

Review of the GRACE-generated Mollie connector Klarna (PayLater) implementation:

---

### 🔴 Critical Issues

#### 1. Unwrap Cascades Could Panic in Production
**File**: `crates/integrations/connector-integration/src/connectors/mollie/transformers.rs`

Multiple `.ok_or()?` chains followed by `.peek().to_string()` could panic if underlying data is malformed:

```rust
// Lines 295-310: Multiple unwrap points
let line1 = address
    .line1
    .as_ref()
    .ok_or(...)?
    .peek()
    .to_string();
```

Consider extracting this into a helper that returns a proper `Result` with context.

---

### 🟠 High Priority Issues

#### 2. Hardcoded "physical" Line Type for Klarna
**File**: Same file, line 348+

```rust
line_type: "physical".to_string(),
```

Klarna requires accurate line types (physical, shipping_fee, discount, etc.) for proper risk assessment. Using "physical" for all payments may cause issues with:
- Digital goods/services
- Mixed carts with shipping
- Discount applications

Consider making this configurable or inferring from product data.

---

### 🟡 Medium Priority

#### 3. Region Field Ignored in Billing Address
**File**: Same file, line 339

```rust
region: None,
```

Mollie's Klarna integration supports region/state. Populating this from `address.state` when available improves address validation.

#### 4. Missing Order Lines Beyond Single Entry
The current implementation creates exactly one order line representing the total payment. For partial captures or multi-item orders, this loses granularity. Document this limitation clearly.

---

### ✅ Positive Aspects

1. **Proper secret wrapping** for sensitive fields (street_and_number, postal_code)
2. **Good error context** with field names in MissingRequiredField errors
3. **Fallback to request.email** when billing email is absent
4. **Clean TryFrom implementation** following existing patterns

---

### 📝 Additional Notes

Since this is GRACE-generated code:
- Verify the generated gRPC tests cover edge cases (missing address fields, invalid countries)
- Ensure the sandbox test actually exercised the redirect flow to completion
- Consider adding a test case for when `line2` is absent

**Recommendation**: Fix the hardcoded "physical" line type and add defensive error handling before merging.

**Review by XyneSpaces — COMMENTED** [review]
## Code Review: Mollie Authorize Flow Implementation

This PR implements the Authorize flow for Mollie connector with Klarna (PayLater) support. The implementation was generated by GRACE automated pipeline.

### Key Findings:

1. **Klarna Payment Method Support**: The new `MolliePaymentMethodData::Klarna` variant with `KlarnaMethodData`, `MollieKlarnaAddress`, and `MollieLineItem` types is well-structured.

2. **Required Field Validation**: The `TryFrom` implementation properly validates required fields (billing address, email, line1, zip, city, country, first_name, last_name) with descriptive error messages.

3. **Address Construction**: The street_and_number construction from line1 and optional line2 is correctly handled.

4. **Order Lines**: Since order line details aren't available in `PaymentsAuthorizeData`, creating a single line representing the full payment amount is a reasonable fallback.

5. **gRPC Test Results**: PASS status confirms the implementation works end-to-end.

### Suggestions:

1. Consider adding unit tests for the `TryFrom` conversion logic
2. The `line_type` is hardcoded as "physical" - verify this is appropriate for all Klarna payment scenarios
3. Consider making the order line description more descriptive when available

Overall, a clean implementation following the existing patterns.

**Comment by XyneSpaces** [comment]
## 📋 PR Review Summary

### Review Status: ✅ Approved

**Key Changes:**
- Implemented Authorize flow for Mollie connector
- Added Klarna (PayLater) payment method support
- GRACE automated pipeline validated implementation

**Findings:**
| Severity | Count | Description |
|----------|-------|-------------|
| 🟡 Question | 1 | Hardcoded "physical" line_type |
| ✅ Positive | 3 | Proper validation, clean types, gRPC tests pass |

**Code Quality:**
- Excellent required field validation with descriptive errors
- Proper address construction from billing details
- Order line fallback for missing details is reasonable

**Action Items:**
1. Verify "physical" line_type is appropriate for all scenarios
2. Consider unit tests for TryFrom conversion logic

Clean implementation following connector integration patterns. gRPC tests confirm functionality.

**Comment by XyneSpaces** [comment]
## Review: Mollie Klarna Implementation

### [should-fix] Line item type hardcoded as "physical"

```rust
line_type: "physical".to_string(),
```

The line type is hardcoded to "physical" but Klarna supports multiple types (shipping_fee, sales_tax, discount, store_credit). If the payment is purely a service/digital product, "physical" may be incorrect. Consider:
1. Deriving from payment metadata if available
2. Using "digital" or a configurable default for non-physical goods

### [blocking] Missing error handling for non-Klarna PayLater methods

The implementation only handles `PayLaterData::KlarnaRedirect` but returns `NotImplemented` for all other PayLater variants. This is correct but ensure the error message is helpful:

```rust
_ => {
    return Err(IntegrationError::NotImplemented(
        "Payment method not yet implemented for Mollie".to_string()
    ))
}
```

Consider including the actual payment method name in the error for easier debugging.

### [should-fix] Address region field ignored

```rust
region: None,
```

The Mollie Klarna address struct supports `region` but it's always set to `None`. If the source address has state/province information, it should be mapped here.

**Review by gopikrishna000 — DISMISSED** [review]

File: crates/integrations/connector-integration/src/connectors/mollie/transformers.rs [hunk 1]
```diff
@@ -223,6 +260,120 @@ impl<T: PaymentMethodDataTypes + std::fmt::Debug + Sync + Send + 'static + Seria
...
+                            field_name: "billing_address.address",
+                            context: Default::default(),
+                        })?;
+
+                let line1 = address
+                    .line1
+                    .as_ref()
+                    .ok_or(IntegrationError::MissingRequiredField {
+                        field_name: "billing_address.line1",
+                        context: Default::default(),
```
Comment (AmitsinghTanwar007) [inline crates/integrations/connector-integration/src/connectors/mollie/transformers.rs]: add more context to all the default

File: crates/integrations/connector-integration/src/connectors/mollie/transformers.rs [hunk 2]
```diff
@@ -223,6 +260,120 @@ impl<T: PaymentMethodDataTypes + std::fmt::Debug + Sync + Send + 'static + Seria
...
+                        },
+                    )?,
+                    email,
+                };
+
+                // Mollie requires at least one order line for Klarna. Order line
+                // details are not available in PaymentsAuthorizeData, so build a
+                // single line representing the full payment amount.
+                let line = MollieLineItem {
+                    line_type: "physical".to_string(),
```
Comment (AmitsinghTanwar007) [inline crates/integrations/connector-integration/src/connectors/mollie/transformers.rs]: why this is hardcoded

**Review by AmitsinghTanwar007 — DISMISSED** [review]

File: crates/integrations/connector-integration/src/connectors/mollie/transformers.rs (hunk 1 above)
Comment (shuklatushar226) [reply crates/integrations/connector-integration/src/connectors/mollie/transformers.rs]: Done in 814337f8c — every `MissingRequiredField` in the Klarna billing-address mapping now sets `IntegrationErrorContext.additional_context` explaining which Klarna requirement is unmet (e.g. "Mollie Klarna requires the street address (line1)..."), so the gRPC error message is self-explanatory instead of a bare "Missing required field".

File: crates/integrations/connector-integration/src/connectors/mollie/transformers.rs (hunk 2 above)
Comment (shuklatushar226) [reply crates/integrations/connector-integration/src/connectors/mollie/transformers.rs]: Mollie requires a `type` on every Klarna order line, but per-product goods type isn't available in `PaymentsAuthorizeData` (we build a single synthetic line for the full amount). Fixed in 814337f8c: replaced the `"physical"` magic string with a typed `MollieLineItemType` enum and documented that `Physical` is the default (most general type Klarna accepts), to be refined once line-item/product-type data is plumbed through.

**Review by AmitsinghTanwar007 — DISMISSED** [review]

**Review by gopikrishna000 — DISMISSED** [review]

**Review by gopikrishna000 — APPROVED** [review]

**Review by AmitsinghTanwar007 — APPROVED** [review]
